### ISSUE TITLE
Batch jobs not migrated after node drain

### ISSUE THREAD

**Issue by wyattanderson** [hashicorp/nomad#7887]
### Nomad version
0.11.1

### Operating system and Environment details
- CentOS 7
- EC2 in AWS
- `docker` and `raw_exec`

### Issue

Since upgrading from 0.8.7 to 0.11.1, we've noticed that batch jobs will occasionally not be migrated after a node is drained. This is separate from `reschedule` or `restart` behavior. Allocations in question look as we would expect in terms of `DesiredStatus`/`DesiredDescription` (`stop`/`alloc is being migrated`).

The only difference I see in the `job status` output (beyond the fact that there's no rescheduled allocation) is that the subsequent evaluation triggered by `node-drain` failed due to exhausted resources (we queue a lot of batch jobs, so this is expected). I don't have enough data right now to know if this is consistent, but we've increased logging to try and capture more details if it happens again.

#### `nomad job status -verbose` for a job that was _not_ migrated as expected

```
ID            = test_runner_slow_1x/dispatch-1588866073-8a7cbe25
Name          = test_runner_slow_1x/dispatch-1588866073-8a7cbe25
Submit Date   = 2020-05-07T11:41:13-04:00
Type          = batch
Priority      = 20
Datacenters   = us-east-1b,us-east-1c,us-east-1d
Namespace     = default
Status        = dead (stopped)
Periodic      = false
Parameterized = false

Summary
Task Group   Queued  Starting  Running  Failed  Complete  Lost
test_runner  0       0         0        0       1         0

Evaluations
ID                                    Priority  Triggered By       Status    Placement Failures
a2cf5ebe-ac43-90c7-b454-830586da64b4  20        node-update        complete  false
19d889fa-31a9-a5c7-0768-ba252b431b07  20        queued-allocs      complete  false
8c9bdc13-574a-b2e7-46b3-a4bd31a100e1  20        node-drain         complete  true
ee7585d5-918f-cf17-8149-0aec605145c2  20        max-plan-attempts  complete  false
d1676153-66ea-959e-c9e9-438c3ff80ec1  20        max-plan-attempts  failed    false

Allocations
ID                                    Eval ID                               Node ID                               Node Name                                    Task Group   Version  Desired  Status    Created                    Modified
ddcba7d9-1ddf-7009-a37d-d8f0e297e99d  ee7585d5-918f-cf17-8149-0aec605145c2  a3a4fac5-4ded-4537-a22a-916a088c39a9  REDACTED  test_runner  0        stop     complete  2020-05-07T12:01:31-04:00  2020-05-07T12:04:38-04:00
```

#### `nomad job status -verbose` for a job that _was_ migrated as expected

```
ID            = test_runner_slow_1x/dispatch-1588850554-fa5aa5f1
Name          = test_runner_slow_1x/dispatch-1588850554-fa5aa5f1
Submit Date   = 2020-05-07T07:22:34-04:00
Type          = batch
Priority      = 20
Datacenters   = us-east-1b,us-east-1c,us-east-1d
Namespace     = default
Status        = dead
Periodic      = false
Parameterized = false

Summary
Task Group   Queued  Starting  Running  Failed  Complete  Lost
test_runner  0       0         0        0       2         0

Evaluations
ID                                    Priority  Triggered By       Status    Placement Failures
e995a9e9-97bf-fb62-ef41-147c873a7119  20        node-update        complete  false
6d021db8-0720-91e6-ec79-2c505dfcf7cc  20        node-update        complete  false
f47db6b4-c2e2-79b2-465f-02ac9895e132  20        node-drain         complete  false
8aa07bf9-da4b-1c5f-d3c5-33efb1dfe1c1  20        max-plan-attempts  complete  false
c3a3bdab-8513-bd39-f93b-ce046a12fbd8  20        max-plan-attempts  failed    false
cfe1cbc5-c574-d97f-3941-4796152d3f9c  20        max-plan-attempts  failed    false
b20352a6-ce0c-93dc-5b4f-f589dc9c5f2a  20        max-plan-attempts  failed    false
95846136-edf8-2519-1280-cca340fcdfec  20        max-plan-attempts  failed    false
1c203b49-e600-d1d1-833a-153b3282ecd0  20        max-plan-attempts  failed    false
dbe20f68-67a8-e079-7e8d-3c5fe80d28d6  20        max-plan-attempts  failed    false
914ee119-b39a-5f3e-d1fa-511aaab32336  20        queued-allocs      failed    false
93aaed3b-581b-fda9-3b3e-df4c08190fc6  20        job-register       complete  true

Allocations
ID                                    Eval ID                               Node ID                               Node Name                                    Task Group   Version  Desired  Status    Created                    Modified
78b63387-8417-e404-8960-8474124f4041  f47db6b4-c2e2-79b2-465f-02ac9895e132  110f667b-c7a9-5e29-12c0-64021e59f555  REDACTED  test_runner  0        run      complete  2020-05-07T07:46:47-04:00  2020-05-07T12:29:28-04:00
c5be4423-a739-86bb-c069-2d5880cd0b18  8aa07bf9-da4b-1c5f-d3c5-33efb1dfe1c1  cad0d245-d658-ef95-07ba-d11e36c8f4d5  REDACTED  test_runner  0        stop     complete  2020-05-07T07:44:53-04:00  2020-05-07T07:46:48-04:00
```

### Reproduction steps

Currently unable to reproduce.

### Job file (if appropriate)

```
job "test_runner_slow_1x" {

    priority = 20

    type = "batch"

    datacenters = ["us-east-1b", "us-east-1c", "us-east-1d"]

    parameterized {
        # redacted
        # meta_required = [...]
        # meta_optional = [...]
    }

    group "test_runner" {
        restart {
            attempts = 0
            mode = "fail"
        }

        reschedule {
            attempts = 0
            unlimited = false
        }

        task "artifact_collector" {
            driver = "docker"

            config {
                image = "redacted"

                args = [
                  # redacted
                ]

                volumes = [
                  # redacted
                ]
            }

            kill_timeout = "30s"

            resources {
                cpu    = 70
                memory = 64
            }
        }

        task "test" {
            driver = "docker"

            leader = true

            resources {
                cpu = 2800
                memory = 3584
            }

            env {
              # redacted
            }

            config {
                image = "redacted"
                command = "/bin/bash"
                args = ["-c", "exec scripts/redacted"]

                volumes = [
                    "test-tmp:/tmp"
                ]
            }
        }
    }
}
```

### Nomad Server logs (if appropriate)

Server logs from the timeframe of the incident don't appear to contain anything interesting:

```
May 07 11:58:39 redacted nomad[11859]: 2020-05-07T11:58:39.238-0400 [WARN]  snapshot: found temporary snapshot: name=1632-312099539-1588093983541.tmp
May 07 11:58:39 redacted nomad[11859]: 2020-05-07T11:58:39.238-0400 [INFO]  snapshot: reaping snapshot: path=/var/lib/nomad/server/raft/snapshots/1640-316712350-1588866176687
May 07 11:58:39 redacted nomad[11859]: 2020-05-07T11:58:39.294-0400 [INFO]  nomad.raft: compacting logs: from=316713196 to=316724084
May 07 11:58:39 redacted nomad[11859]: 2020-05-07T11:58:39.321-0400 [INFO]  nomad.raft: snapshot complete up to: index=316734056
May 07 12:05:46 redacted nomad[11859]: 2020-05-07T12:05:46.365-0400 [INFO]  nomad.raft: starting snapshot up to: index=316745850
May 07 12:05:46 redacted nomad[11859]: 2020-05-07T12:05:46.365-0400 [INFO]  snapshot: creating new snapshot: path=/var/lib/nomad/server/raft/snapshots/1640-316745850-1588867546365.tmp
May 07 12:05:52 redacted nomad[11859]: 2020-05-07T12:05:52.261-0400 [WARN]  snapshot: found temporary snapshot: name=1632-312099539-1588093983541.tmp
May 07 12:05:52 redacted nomad[11859]: 2020-05-07T12:05:52.261-0400 [INFO]  snapshot: reaping snapshot: path=/var/lib/nomad/server/raft/snapshots/1640-316723167-1588866725562
May 07 12:05:52 redacted nomad[11859]: 2020-05-07T12:05:52.314-0400 [INFO]  nomad.raft: compacting logs: from=316724085 to=316735808
May 07 12:05:52 redacted nomad[11859]: 2020-05-07T12:05:52.333-0400 [INFO]  nomad.raft: snapshot complete up to: index=316745850
May 07 12:14:00 redacted nomad[11859]: 2020-05-07T12:14:00.187-0400 [INFO]  nomad.drain: node completed draining at index: node_id=9ac81855-855d-c09a-8546-c1ad3117dd94 index=316747276
May 07 12:14:14 redacted nomad[11859]: 2020-05-07T12:14:14.826-0400 [WARN]  nomad.heartbeat: node TTL expired: node_id=a8e14245-6645-cdb2-c51d-3902636a39d4
May 07 12:14:16 redacted nomad[11859]: 2020-05-07T12:14:16.161-0400 [WARN]  nomad.heartbeat: node TTL expired: node_id=201d427a-436d-cbff-a4c7-31018c4475be
May 07 12:14:16 redacted nomad[11859]: 2020-05-07T12:14:16.311-0400 [WARN]  nomad.heartbeat: node TTL expired: node_id=a3a4fac5-4ded-4537-a22a-916a088c39a9
May 07 12:14:16 redacted nomad[11859]: 2020-05-07T12:14:16.359-0400 [WARN]  nomad.heartbeat: node TTL expired: node_id=52b34d42-c6a3-84e9-83f0-d60889a79ebb
May 07 12:14:18 redacted nomad[11859]: 2020-05-07T12:14:18.231-0400 [WARN]  nomad.heartbeat: node TTL expired: node_id=69db13bf-f040-2160-2882-eb7aab7581da
May 07 12:14:19 redacted nomad[11859]: 2020-05-07T12:14:19.088-0400 [WARN]  nomad.heartbeat: node TTL expired: node_id=14cd20e2-d002-8e56-c8c3-0a693f301e90
May 07 12:14:19 redacted nomad[11859]: 2020-05-07T12:14:19.175-0400 [WARN]  nomad.heartbeat: node TTL expired: node_id=d9332dd9-a2be-fb60-ab81-02e42dbd78ee
May 07 12:14:24 redacted nomad[11859]: 2020-05-07T12:14:24.848-0400 [WARN]  nomad.heartbeat: node TTL expired: node_id=6da6bf18-1209-0f27-ae6b-f2dbc3129d4f
May 07 12:14:25 redacted nomad[11859]: 2020-05-07T12:14:25.774-0400 [WARN]  nomad.heartbeat: node TTL expired: node_id=7be0dc49-0aa6-e2c8-5c50-3bab9b71a7f8
May 07 12:14:25 redacted nomad[11859]: 2020-05-07T12:14:25.953-0400 [WARN]  nomad.heartbeat: node TTL expired: node_id=294aa898-82a6-9b41-6e9c-46a92cfab238
May 07 12:14:26 redacted nomad[11859]: 2020-05-07T12:14:26.936-0400 [WARN]  nomad.heartbeat: node TTL expired: node_id=10a078e0-9ed9-d222-82e8-cf86e0507d71
May 07 12:14:27 redacted nomad[11859]: 2020-05-07T12:14:27.139-0400 [WARN]  nomad.heartbeat: node TTL expired: node_id=5410c6e4-3b6d-1384-dc40-f1e00f1b7e2a
May 07 12:14:27 redacted nomad[11859]: 2020-05-07T12:14:27.759-0400 [WARN]  nomad.heartbeat: node TTL expired: node_id=936d8434-1876-29e9-087a-f55a88ad7879
May 07 12:14:28 redacted nomad[11859]: 2020-05-07T12:14:28.036-0400 [WARN]  nomad.heartbeat: node TTL expired: node_id=3bd507bf-7077-9eae-bcfe-e313933c4ba5
```

TTL expirations are expected; nodes were drained and terminated in reaction to demand. It is interesting that there's only one `node completed draining` message; these appear to be sporadic relative to the frequency with which we drain and terminate nodes.

**Comment by tgross** [comment]
Hi @wyattanderson! 

Thanks for opening this issue with as much detail as you could! There are a couple other open questions around seemingly incorrect Node drain behaviors so I'm going to set aside some time to dive deep on them soon. Just one clarifying question:

> Since upgrading from 0.8.7 to 0.11.1, we've noticed that batch jobs will occasionally not be migrated after a node is drained.

Normally we recommend upgrading one major version at a time. Is there any chance that the problem you're seeing is isolated to jobs that were created under 0.8.7 and then were migrated up to 0.11.1 without a new version being run?

**Comment by wyattanderson** [comment]
Thanks for taking a look!

> Normally we recommend upgrading one major version at a time. Is there any chance that the problem you're seeing is isolated to jobs that were created under 0.8.7 and then were migrated up to 0.11.1 without a new version being run?

Hmm. These jobs were dispatched well after the upgrade, but the parameterized job may have been registered before the upgrade. We recently reloaded the job configurations and have yet to see the issue again, so it's possible it's related, but difficult to say for sure.

Possibly worth noting, our automation was draining nodes with a 1 minute deadline; we've adjusted that to 10 minutes now to try and prevent this from happening (as most of our jobs will terminate well within that window), so I'm unsure if and when we'll run into this issue again.
